File: clients/js/capture-client.js
/**
 * CaptureClient — JS helper for the mock proxy's capture-session API.
 *
 * Lets automated tests mark a START, exercise the app through the proxy,
 * mark an END, and assert on the exact requests the app sent in that window:
 *
 *   const { CaptureClient } = require("./capture-client");
 *   const client = new CaptureClient();
 *   await client.start({ name: "login-test" });
 *   // ... drive the app ...
 *   const { requests } = await client.stop();
 *   const login = requests.find((r) => r.path === "/api/login");
 *   expect(login.requestBody).toEqual({ user: "u", pass: "p" });
 *
 * Standalone on purpose: zero dependencies (node:http only) so this file can
 * be copied into any QA repo. The port autodetection below intentionally
 * duplicates scripts/cli.js — keeping the client droppable beats sharing.
 */
const http = require("http");

const PREFERRED_PORT = 8888; // proxy's preferred port (often already taken)
const PORT_SCAN = 20; // matches the proxy's own fallback scan range

/** Probe a port for OUR admin server (another proxy on 8888 won't match). */
function _probe(host, port) {
  return new Promise((resolve) => {
    const req = http.request(
      { hostname: host, port, path: "/__admin/health", method: "GET", timeout: 500 },
      (res) => {
        let data = "";
        res.on("data", (ch) => (data += ch));
        res.on("end", () => {
          try {
            const j = JSON.parse(data);
            resolve(!!j && j.status === "ok" && Array.isArray(j.instances));
          } catch {
            resolve(false);
          }
        });
      }
    );
    req.on("error", () => resolve(false));
    req.on("timeout", () => {
      req.destroy();
      resolve(false);
    });
    req.end();
  });
}

class CaptureClient {
  /**
   * @param {{ host?: string, port?: number }} [opts] — explicit port skips
   *   autodetection (recommended in hermetic tests). Defaults honor the
   *   MOCK_HOST / MOCK_PORT environment variables.
   */
  constructor({ host, port } = {}) {
    this.host = host || process.env.MOCK_HOST || "localhost";
    this.port =
      port || (process.env.MOCK_PORT ? parseInt(process.env.MOCK_PORT, 10) : null);
    this.sessionId = null;
  }

  async _resolvePort() {
    if (this.port) return this.port;
    for (let p = PREFERRED_PORT; p <= PREFERRED_PORT + PORT_SCAN; p++) {
      if (await _probe(this.host, p)) return (this.port = p);
    }
    throw new Error(
      `CaptureClient: no mock proxy found on ${this.host}:${PREFERRED_PORT}-${
        PREFERRED_PORT + PORT_SCAN
      } (is the server running? set MOCK_PORT to override)`
    );
  }

  async _request(method, path, body) {
    const port = await this._resolvePort();
    const payload = body ? JSON.stringify(body) : null;
    const { status, data } = await new Promise((resolve, reject) => {
      const headers = { "Content-Type": "application/json" };
      if (payload) headers["Content-Length"] = Buffer.byteLength(payload);
      const req = http.request(
        { hostname: this.host, port, path: `/__admin${path}`, method, headers },
        (res) => {
          let raw = "";
          res.on("data", (ch) => (raw += ch));
          res.on("end", () => resolve({ status: res.statusCode, data: raw }));
        }
      );
      req.on("error", reject);
      if (payload) req.write(payload);
      req.end();
    });
    let parsed;
    try {
      parsed = JSON.parse(data);
    } catch {
      parsed = data;
    }
    if (status < 200 || status >= 300) {
      const msg = (parsed && parsed.error) || data || `HTTP ${status}`;
      throw new Error(`CaptureClient: ${method} ${path} failed (${status}): ${msg}`);
    }
    return parsed;
  }

  /**
   * Start a capture session. The id is returned and remembered on the
   * instance, so subsequent calls may omit it.
   * @param {{ name?: string, instanceId?: string }} [opts] — instanceId
   *   restricts the capture to one configured backend instance.
   * @returns {Promise<string>} sessionId
   */
  async start({ name, instanceId } = {}) {
    const res = await this._request("POST", "/capture/start", { name, instanceId });
    this.sessionId = res.sessionId;
    return res.sessionId;
  }

  /**
   * Stop the session and return everything captured in the window.
   * Idempotent. Requests are chronological (oldest first).
   * @returns {Promise<{ sessionId, status, count, droppedCount, requests }>}
   */
  async stop(sessionId = this.sessionId) {
    if (!sessionId)
      throw new Error("CaptureClient: no active session (call start first)");
    return this._request("POST", "/capture/stop", { sessionId });
  }

  /**
   * Query a session's captured requests with optional filters:
   * method (exact, case-insensitive), path (exact pathname),
   * pathPrefix, instanceId, source ("mock"|"proxy"|"intercept"|"server-off").
   * @returns {Promise<object[]>}
   */
  async getRequests(sessionId = this.sessionId, filters = {}) {
    if (!sessionId)
      throw new Error("CaptureClient: no active session (call start first)");
    const qs = new URLSearchParams(
      Object.entries(filters).filter(([, v]) => v !== undefined && v !== null)
    ).toString();
    const res = await this._request(
      "GET",
      `/capture/${encodeURIComponent(sessionId)}/requests${qs ? `?${qs}` : ""}`
    );
    return res.requests;
  }

  /** Session metadata (status, count, droppedCount) without the entries. */
  async getSession(sessionId = this.sessionId) {
    if (!sessionId)
      throw new Error("CaptureClient: no active session (call start first)");
    const res = await this._request("GET", `/capture/${encodeURIComponent(sessionId)}`);
    return res.session;
  }

  /** Delete a session server-side (e.g., in afterEach cleanup). */
  async delete(sessionId = this.sessionId) {
    if (!sessionId)
      throw new Error("CaptureClient: no active session (call start first)");
    await this._request("DELETE", `/capture/${encodeURIComponent(sessionId)}`);
    if (sessionId === this.sessionId) this.sessionId = null;
  }
}

module.exports = { CaptureClient };

File: clients/python/capture_client.py
"""CaptureClient -- Python helper for the mock proxy's capture-session API.

Lets automated tests mark a START, exercise the app through the proxy, mark
an END, and assert on the exact requests the app sent in that window:

    from capture_client import capture

    with capture(name="login-test") as session:
        app.login("user", "pass")  # drive the app through the proxy

    login = next(r for r in session.requests if r["path"] == "/api/login")
    assert login["requestBody"] == {"user": "user", "pass": "pass"}

Standalone on purpose: stdlib only, so this file can be copied into any QA
repo. The port autodetection below intentionally duplicates the project's
scripts/cli.js -- keeping the client droppable beats sharing code.
"""

import json
import os
import urllib.error
import urllib.parse
import urllib.request
from contextlib import contextmanager

PREFERRED_PORT = 8888  # proxy's preferred port (often already taken)
PORT_SCAN = 20  # matches the proxy's own fallback scan range


class CaptureError(Exception):
    """Raised when the capture API returns an error response."""


def _probe(host, port):
    """Probe a port for OUR admin server (another proxy on 8888 won't match)."""
    try:
        with urllib.request.urlopen(
            f"http://{host}:{port}/__admin/health", timeout=0.5
        ) as res:
            data = json.loads(res.read().decode("utf-8"))
            return data.get("status") == "ok" and isinstance(data.get("instances"), list)
    except Exception:
        return False


class CaptureClient:
    def __init__(self, host=None, port=None):
        """Explicit `port` skips autodetection (recommended in hermetic tests).
        Defaults honor the MOCK_HOST / MOCK_PORT environment variables."""
        self.host = host or os.environ.get("MOCK_HOST", "localhost")
        env_port = os.environ.get("MOCK_PORT")
        self.port = port or (int(env_port) if env_port else None)
        self.session_id = None

    def _resolve_port(self):
        if self.port:
            return self.port
        for p in range(PREFERRED_PORT, PREFERRED_PORT + PORT_SCAN + 1):
            if _probe(self.host, p):
                self.port = p
                return p
        raise CaptureError(
            f"no mock proxy found on {self.host}:{PREFERRED_PORT}-"
            f"{PREFERRED_PORT + PORT_SCAN} (is the server running? "
            "set MOCK_PORT to override)"
        )

    def _request(self, method, path, body=None):
        port = self._resolve_port()
        url = f"http://{self.host}:{port}/__admin{path}"
        payload = json.dumps(body).encode("utf-8") if body is not None else None
        req = urllib.request.Request(
            url, data=payload, method=method,
            headers={"Content-Type": "application/json"},
        )
        try:
            with urllib.request.urlopen(req, timeout=10) as res:
                return json.loads(res.read().decode("utf-8"))
        except urllib.error.HTTPError as e:
            try:
                detail = json.loads(e.read().decode("utf-8")).get("error", "")
            except Exception:
                detail = ""
            raise CaptureError(
                f"{method} {path} failed ({e.code}): {detail or e.reason}"
            ) from None

    def start(self, name=None, instance_id=None):
        """Start a capture session; returns and remembers the session id.
        `instance_id` restricts the capture to one configured backend."""
        body = {}
        if name:
            body["name"] = name
        if instance_id:
            body["instanceId"] = instance_id
        res = self._request("POST", "/capture/start", body)
        self.session_id = res["sessionId"]
        return self.session_id

    def stop(self, session_id=None):
        """Stop the session; returns the full payload including `requests`
        (chronological, oldest first). Idempotent."""
        sid = session_id or self._require_session()
        return self._request("POST", "/capture/stop", {"sessionId": sid})

    def get_requests(self, session_id=None, method=None, path=None,
                     path_prefix=None, instance_id=None, source=None):
        """Captured requests with optional filters: method (exact,
        case-insensitive), path (exact pathname), path_prefix, instance_id,
        source ("mock" | "proxy" | "intercept" | "server-off")."""
        sid = session_id or self._require_session()
        params = {
            "method": method, "path": path, "pathPrefix": path_prefix,
            "instanceId": instance_id, "source": source,
        }
        qs = urllib.parse.urlencode({k: v for k, v in params.items() if v})
        suffix = f"?{qs}" if qs else ""
        res = self._request(
            "GET", f"/capture/{urllib.parse.quote(sid)}/requests{suffix}"
        )
        return res["requests"]

    def get_session(self, session_id=None):
        """Session metadata (status, count, droppedCount) without entries."""
        sid = session_id or self._require_session()
        return self._request("GET", f"/capture/{urllib.parse.quote(sid)}")["session"]

    def delete(self, session_id=None):
        """Delete a session server-side (e.g., in test teardown)."""
        sid = session_id or self._require_session()
        self._request("DELETE", f"/capture/{urllib.parse.quote(sid)}")
        if sid == self.session_id:
            self.session_id = None

    def _require_session(self):
        if not self.session_id:
            raise CaptureError("no active session (call start first)")
        return self.session_id


class _CaptureSession:
    """Result holder for the `capture` context manager."""

    def __init__(self, client):
        self.client = client
        self.session_id = None
        self.requests = None
        self.dropped_count = 0

    def get_requests(self, **filters):
        return self.client.get_requests(self.session_id, **filters)


@contextmanager
def capture(name=None, instance_id=None, host=None, port=None):
    """Capture everything the app sends through the proxy inside the block.

    with capture("checkout") as session:
        ...drive the app...
    assert session.requests[0]["requestBody"] == {...}
    """
    client = CaptureClient(host=host, port=port)
    session = _CaptureSession(client)
    session.session_id = client.start(name=name, instance_id=instance_id)
    try:
        yield session
    finally:
        result = client.stop(session.session_id)
        session.requests = result["requests"]
        session.dropped_count = result.get("droppedCount", 0)
